File: Sources/XDecodeCore/ZipDecodeCoordinator.swift
import Darwin
import Foundation
import ZIPFoundation

public typealias ArchiveEntryFormatResolver = @Sendable (URL) async -> LogFormat?

public protocol ZipOutputPublicationTracking: Sendable {
    func preparePublication(stagingDirectory: URL, destinationDirectory: URL) async throws
    func cancelPublication(stagingDirectory: URL) async
}

public actor ZipDecodeCoordinator {
    private struct ArchiveItem {
        let entry: Entry
        let relativePath: String
        let format: LogFormat?
    }

    private let fileManager: FileManager
    private let decoderResolver: DecoderResolver
    private let entryFormatResolver: ArchiveEntryFormatResolver
    private let publicationTracker: (any ZipOutputPublicationTracking)?
    private let maximumArchiveEntryCount = 1_000
    private let maximumLogCount = 100
    private let maximumEntrySize = DecodeLimits.maximumInputFileSize
    private let maximumTotalSize: UInt64 = 1024 * 1024 * 1024

    public init(
        fileManager: FileManager = .default,
        decoderResolver: @escaping DecoderResolver,
        entryFormatResolver: @escaping ArchiveEntryFormatResolver = {
            try? LogFormat.detect(from: $0)
        },
        publicationTracker: (any ZipOutputPublicationTracking)? = nil
    ) {
        self.fileManager = fileManager
        self.decoderResolver = decoderResolver
        self.entryFormatResolver = entryFormatResolver
        self.publicationTracker = publicationTracker
    }

    public func decode(_ request: DecodeRequest) async -> DecodeResult {
        let sourceURL = request.sourceURL.standardizedFileURL
        var temporaryDirectory: URL?

        do {
            guard request.format == .zip else {
                throw DecodeError.unsupportedFormat(request.format.rawValue)
            }
            guard fileManager.fileExists(atPath: sourceURL.path) else {
                throw DecodeError.fileOperation("源 ZIP 不存在")
            }
            try DecodeLimits.validateInputFile(at: sourceURL, description: "源 ZIP")

            let archive: Archive
            do {
                archive = try Archive(url: sourceURL, accessMode: .read)
            } catch {
                throw DecodeError.malformed("ZIP 无法读取：\(error.localizedDescription)")
            }

            let items = try await inspectEntries(in: archive, sourceURL: sourceURL)
            let logCount = items.filter { $0.format != nil }.count
            guard logCount > 0 else {
                return DecodeResult(
                    request: request,
                    state: .skipped,
                    outputURL: nil,
                    message: "ZIP 中没有符合当前规则的 Xlog、MX 或 Logan 日志；未生成输出目录，源 ZIP 已保留",
                    sourceDeleted: false
                )
            }
            guard logCount <= maximumLogCount else {
                throw DecodeError.decodingFailed("ZIP 中日志数量超过 \(maximumLogCount) 个")
            }

            let parentDirectory = sourceURL.deletingLastPathComponent()
            let stagingDirectory = parentDirectory.appendingPathComponent(
                ".xdecode-zip-\(UUID().uuidString).tmp",
                isDirectory: true
            )
            try fileManager.createDirectory(
                at: stagingDirectory,
                withIntermediateDirectories: false
            )
            temporaryDirectory = stagingDirectory

            for item in items where item.format == nil {
                try extractPreservedItem(item, from: archive, to: stagingDirectory)
            }

            var decodedCount = 0
            var decodedOutputSize = 0
            var failedLogNames = [String]()
            for item in items {
                guard let format = item.format else { continue }
                let sourceEntryURL = sourceURL.deletingLastPathComponent()
                    .appendingPathComponent(item.relativePath, isDirectory: false)
                let payload = try extractLogData(item, from: archive)
                let entryRequest = try DecodeRequest(
                    sourceURL: sourceEntryURL,
                    format: format,
                    origin: request.origin,
                    requestedAt: request.requestedAt
                )

                var decodedOutputURL: URL?
                do {
                    let decoder = try await decoderResolver(entryRequest)
                    let decoded = try decoder.decode(payload, sourceURL: sourceEntryURL)
                    guard decoded.isComplete else {
                        throw DecodeError.decodingFailed(decoded.diagnostic ?? "日志未完整解密")
                    }
                    guard !decoded.data.isEmpty else { throw DecodeError.emptyOutput }
                    try DecodeLimits.validateDecompressedOutputSize(
                        currentSize: decodedOutputSize,
                        appending: decoded.data.count
                    )

                    let outputURL = uniqueDecodedOutput(
                        for: item.relativePath,
                        in: stagingDirectory
                    )
                    decodedOutputURL = outputURL
                    try write(decoded.data, to: outputURL)
                    decodedOutputSize += decoded.data.count
                    decodedCount += 1
                } catch {
                    if let decodedOutputURL {
                        try? fileManager.removeItem(at: decodedOutputURL)
                    }
                    let originalURL = stagingDirectory.appendingPathComponent(
                        item.relativePath,
                        isDirectory: false
                    )
                    do {
                        try write(payload, to: originalURL)
                    } catch {
                        throw DecodeError.fileOperation(
                            "无法保留原始日志 \(item.relativePath)：\(error.localizedDescription)"
                        )
                    }
                    failedLogNames.append((item.relativePath as NSString).lastPathComponent)
                }
            }

            let outputDirectory = try await moveToUniqueOutputDirectory(
                temporaryDirectory: stagingDirectory,
                sourceURL: sourceURL
            )
            temporaryDirectory = nil

            let failedCount = failedLogNames.count
            let state: DecodeState
            let message: String
            if failedCount == 0 {
                state = .completed
                message = "ZIP 批量解密成功：\(decodedCount)/\(logCount) 个日志已生成 .log；源 ZIP 已保留"
            } else if decodedCount == 0 {
                state = .failed
                message = "ZIP 批量解密失败：0/\(logCount) 个日志成功；全部原始日志已保留在输出目录；源 ZIP 已保留"
            } else {
                state = .partiallyCompleted
                let names = failedLogNames.prefix(5).joined(separator: "、")
                let remaining = failedCount > 5 ? "等 \(failedCount) 个文件" : ""
                message = "ZIP 批量部分成功：成功 \(decodedCount) 个，失败 \(failedCount) 个；\(names)\(remaining) 已保留原始日志；源 ZIP 已保留"
            }
            return DecodeResult(
                request: request,
                state: state,
                outputURL: outputDirectory,
                message: message,
                sourceDeleted: false
            )
        } catch {
            if let temporaryDirectory {
                await publicationTracker?.cancelPublication(stagingDirectory: temporaryDirectory)
                try? fileManager.removeItem(at: temporaryDirectory)
            }
            return DecodeResult(
                request: request,
                state: .failed,
                outputURL: nil,
                message: "\(error.localizedDescription)；未生成输出目录，源 ZIP 已保留",
                sourceDeleted: false
            )
        }
    }

    private func inspectEntries(in archive: Archive, sourceURL: URL) async throws -> [ArchiveItem] {
        var items = [ArchiveItem]()
        var seenPaths = Set<String>()
        var totalSize: UInt64 = 0
        var archiveEntryCount = 0

        for entry in archive {
            archiveEntryCount += 1
            guard archiveEntryCount <= maximumArchiveEntryCount else {
                throw DecodeError.decodingFailed("ZIP 条目总数超过 \(maximumArchiveEntryCount) 个")
            }
            let relativePath = try validatedRelativePath(entry.path)
            let comparisonPath = relativePath.lowercased()
            guard seenPaths.insert(comparisonPath).inserted else {
                throw DecodeError.decodingFailed("ZIP 包含重复路径：\(relativePath)")
            }
            if isMetadataPath(relativePath) { continue }

            if entry.type != .directory {
                guard entry.uncompressedSize <= maximumEntrySize else {
                    throw DecodeError.decodingFailed("ZIP 条目超过 500 MB：\(relativePath)")
                }
                let (newTotal, overflow) = totalSize.addingReportingOverflow(entry.uncompressedSize)
                guard !overflow, newTotal <= maximumTotalSize else {
                    throw DecodeError.decodingFailed("ZIP 解压后总大小超过 1 GB")
                }
                totalSize = newTotal
            }

            var format: LogFormat?
            if entry.type == .file {
                let virtualURL = sourceURL.deletingLastPathComponent()
                    .appendingPathComponent(relativePath, isDirectory: false)
                let detected = await entryFormatResolver(virtualURL)
                if let detected, [.xlog, .mx, .logan].contains(detected) {
                    format = detected
                }
            }
            items.append(ArchiveItem(entry: entry, relativePath: relativePath, format: format))
        }
        return items
    }

    private func extractPreservedItem(
        _ item: ArchiveItem,
        from archive: Archive,
        to stagingDirectory: URL
    ) throws {
        let destinationURL = stagingDirectory.appendingPathComponent(
            item.relativePath,
            isDirectory: item.entry.type == .directory
        )
        do {
            let checksum = try archive.extract(
                item.entry,
                to: destinationURL,
                bufferSize: 64 * 1024
            )
            guard checksum == item.entry.checksum else {
                throw DecodeError.decodingFailed("\(item.relativePath) ZIP CRC 校验失败")
            }
            if item.entry.type == .file {
                let handle = try FileHandle(forWritingTo: destinationURL)
                try handle.synchronize()
                try handle.close()
            }
        } catch let error as DecodeError {
            throw error
        } catch {
            throw DecodeError.decodingFailed(
                "\(item.relativePath) 解压失败：\(error.localizedDescription)"
            )
        }
    }

    private func extractLogData(_ item: ArchiveItem, from archive: Archive) throws -> Data {
        let entry = item.entry
        let relativePath = item.relativePath
        let entrySizeLimit = maximumEntrySize
        var payload = Data()
        payload.reserveCapacity(Int(entry.uncompressedSize))
        let checksum: CRC32
        do {
            checksum = try archive.extract(entry, bufferSize: 64 * 1024) { chunk in
                let nextSize = UInt64(payload.count) + UInt64(chunk.count)
                guard nextSize <= entrySizeLimit else {
                    throw DecodeError.decodingFailed(
                        "\(relativePath) 解压后超过大小限制"
                    )
                }
                payload.append(chunk)
            }
        } catch let error as DecodeError {
            throw error
        } catch {
            throw DecodeError.decodingFailed(
                "\(relativePath) 解压失败：\(error.localizedDescription)"
            )
        }
        guard checksum == entry.checksum else {
            throw DecodeError.decodingFailed("\(relativePath) ZIP CRC 校验失败")
        }
        guard UInt64(payload.count) == entry.uncompressedSize else {
            throw DecodeError.decodingFailed("\(relativePath) 解压后的大小不完整")
        }
        return payload
    }

    private func uniqueDecodedOutput(for relativePath: String, in directory: URL) -> URL {
        let path = relativePath as NSString
        let parent = path.deletingLastPathComponent
        let baseName = (path.lastPathComponent as NSString).deletingPathExtension
        var index = 0

        while true {
            let suffix = index == 0 ? "" : "-\(index)"
            let fileName = "\(baseName)\(suffix).log"
            let candidatePath = parent.isEmpty ? fileName : "\(parent)/\(fileName)"
            let candidate = directory.appendingPathComponent(candidatePath, isDirectory: false)
            if !pathExists(candidate) { return candidate }
            index += 1
        }
    }

    private func write(_ data: Data, to url: URL) throws {
        try fileManager.createDirectory(
            at: url.deletingLastPathComponent(),
            withIntermediateDirectories: true
        )
        try data.write(to: url, options: .withoutOverwriting)
        let handle = try FileHandle(forWritingTo: url)
        try handle.synchronize()
        try handle.close()
    }

    private func pathExists(_ url: URL) -> Bool {
        var fileStatus = stat()
        return url.path.withCString { lstat($0, &fileStatus) } == 0
    }

    private func moveToUniqueOutputDirectory(
        temporaryDirectory: URL,
        sourceURL: URL
    ) async throws -> URL {
        let parent = sourceURL.deletingLastPathComponent()
        let baseName = sourceURL.deletingPathExtension().lastPathComponent
        var index = 0

        while true {
            let suffix = index == 0 ? "" : "-\(index)"
            let candidate = parent.appendingPathComponent("\(baseName)\(suffix)", isDirectory: true)
            try await publicationTracker?.preparePublication(
                stagingDirectory: temporaryDirectory,
                destinationDirectory: candidate
            )
            let result = temporaryDirectory.path.withCString { sourcePath in
                candidate.path.withCString { destinationPath in
                    renamex_np(sourcePath, destinationPath, UInt32(RENAME_EXCL))
                }
            }
            if result == 0 { return candidate }

            let renameError = errno
            if renameError == EEXIST {
                index += 1
            } else {
                throw DecodeError.fileOperation(String(cString: strerror(renameError)))
            }
        }
    }

    private func validatedRelativePath(_ rawPath: String) throws -> String {
        let normalized = rawPath.replacingOccurrences(of: "\\", with: "/")
        guard !normalized.isEmpty,
              !normalized.hasPrefix("/"),
              !normalized.contains("\0") else {
            throw DecodeError.decodingFailed("ZIP 包含不安全路径：\(rawPath)")
        }

        var safeComponents = [Substring]()
        for component in normalized.split(separator: "/", omittingEmptySubsequences: false) {
            if component.isEmpty || component == "." { continue }
            guard component != ".." else {
                throw DecodeError.decodingFailed("ZIP 包含路径穿越条目：\(rawPath)")
            }
            safeComponents.append(component)
        }
        guard let first = safeComponents.first,
              !first.hasSuffix(":") else {
            throw DecodeError.decodingFailed("ZIP 包含不安全路径：\(rawPath)")
        }
        return safeComponents.joined(separator: "/")
    }

    private func isMetadataPath(_ relativePath: String) -> Bool {
        let components = relativePath.split(separator: "/")
        return components.contains("__MACOSX")
            || components.last.map { $0.hasPrefix("._") } == true
    }
}

File: Windows/src/XDecode.Core/ZipDecodeCoordinator.cs
using System.IO.Compression;
using System.IO.Hashing;

namespace XDecode.Core;

public delegate ValueTask<LogFormat?> ArchiveEntryFormatResolver(string path, CancellationToken cancellationToken);

public interface IZipOutputPublicationTracker
{
    ValueTask PreparePublicationAsync(string stagingDirectory, string destinationDirectory, CancellationToken cancellationToken);
    ValueTask CancelPublicationAsync(string stagingDirectory, CancellationToken cancellationToken);
}

public sealed class ZipDecodeCoordinator(
    DecoderResolver decoderResolver,
    ArchiveEntryFormatResolver? entryFormatResolver = null,
    IZipOutputPublicationTracker? publicationTracker = null)
{
    private const int MaximumArchiveEntryCount = 1_000;
    private const int MaximumLogCount = 100;
    private const long MaximumEntrySize = DecodeLimits.MaximumInputFileSize;
    private const long MaximumTotalSize = 1024L * 1024 * 1024;
    private readonly ArchiveEntryFormatResolver _entryFormatResolver =
        entryFormatResolver ?? DefaultEntryFormatResolver;

    public async Task<DecodeResult> DecodeAsync(
        DecodeRequest request,
        CancellationToken cancellationToken = default)
    {
        var sourcePath = Path.GetFullPath(request.SourcePath);
        string? stagingDirectory = null;
        try
        {
            if (request.Format != LogFormat.Zip) throw DecodeException.UnsupportedFormat(request.Format.ToString());
            if (!File.Exists(sourcePath)) throw DecodeException.FileOperation("源 ZIP 不存在");
            DecodeLimits.ValidateInputFile(sourcePath, "源 ZIP");

            using var file = new FileStream(sourcePath, FileMode.Open, FileAccess.Read, FileShare.Read);
            using var archive = new ZipArchive(file, ZipArchiveMode.Read, leaveOpen: false);
            var items = await InspectEntriesAsync(archive, sourcePath, cancellationToken).ConfigureAwait(false);
            var logCount = items.Count(value => value.Format is not null);
            if (logCount == 0)
            {
                return new DecodeResult(
                    request, DecodeState.Skipped, null,
                    "ZIP 中没有符合当前规则的 Xlog、MX 或 Logan 日志；未生成输出目录，源 ZIP 已保留",
                    false);
            }
            if (logCount > MaximumLogCount)
                throw DecodeException.DecodingFailed($"ZIP 中日志数量超过 {MaximumLogCount} 个");

            var parent = Path.GetDirectoryName(sourcePath)
                ?? throw DecodeException.FileOperation("无法确定源 ZIP 目录");
            stagingDirectory = Path.Combine(parent, $".xdecode-zip-{Guid.NewGuid():N}.tmp");
            Directory.CreateDirectory(stagingDirectory);
            File.SetAttributes(stagingDirectory, FileAttributes.Directory | FileAttributes.Hidden);

            foreach (var item in items.Where(value => value.Format is null))
                await ExtractPreservedItemAsync(item, stagingDirectory, cancellationToken).ConfigureAwait(false);

            var decodedCount = 0;
            long decodedOutputSize = 0;
            var failedLogNames = new List<string>();
            foreach (var item in items.Where(value => value.Format is not null))
            {
                var virtualPath = Path.Combine(parent, item.RelativePath.Replace('/', Path.DirectorySeparatorChar));
                var payload = await ExtractDataAsync(item, cancellationToken).ConfigureAwait(false);
                var entryRequest = new DecodeRequest(
                    Guid.NewGuid(), virtualPath, item.Format!.Value, request.Origin, request.RequestedAt);
                string? decodedOutputPath = null;
                try
                {
                    var decoder = await decoderResolver(entryRequest, cancellationToken).ConfigureAwait(false);
                    var decoded = decoder.Decode(payload, virtualPath);
                    if (!decoded.IsComplete)
                        throw DecodeException.DecodingFailed(decoded.Diagnostic ?? "日志未完整解密");
                    if (decoded.Data.Length == 0) throw DecodeException.EmptyOutput();
                    DecodeLimits.ValidateOutputSize(decodedOutputSize, decoded.Data.Length);
                    decodedOutputPath = UniqueDecodedOutput(item.RelativePath, stagingDirectory);
                    await WriteExclusiveAsync(decoded.Data, decodedOutputPath, cancellationToken).ConfigureAwait(false);
                    decodedOutputSize += decoded.Data.Length;
                    decodedCount++;
                }
                catch (Exception exception) when (exception is not OperationCanceledException)
                {
                    TryDelete(decodedOutputPath);
                    var originalPath = Path.Combine(
                        stagingDirectory, item.RelativePath.Replace('/', Path.DirectorySeparatorChar));
                    await WriteExclusiveAsync(payload, originalPath, cancellationToken).ConfigureAwait(false);
                    failedLogNames.Add(Path.GetFileName(item.RelativePath));
                }
            }

            var outputDirectory = await MoveToUniqueOutputDirectoryAsync(
                stagingDirectory, sourcePath, cancellationToken).ConfigureAwait(false);
            stagingDirectory = null;
            var failedCount = failedLogNames.Count;
            if (failedCount == 0)
            {
                return new DecodeResult(
                    request, DecodeState.Completed, outputDirectory,
                    $"ZIP 批量解密成功：{decodedCount}/{logCount} 个日志已生成 .log；源 ZIP 已保留", false);
            }
            if (decodedCount == 0)
            {
                return new DecodeResult(
                    request, DecodeState.Failed, outputDirectory,
                    $"ZIP 批量解密失败：0/{logCount} 个日志成功；全部原始日志已保留在输出目录；源 ZIP 已保留", false);
            }
            var names = string.Join('、', failedLogNames.Take(5));
            var remaining = failedCount > 5 ? $"等 {failedCount} 个文件" : "";
            return new DecodeResult(
                request, DecodeState.PartiallyCompleted, outputDirectory,
                $"ZIP 批量部分成功：成功 {decodedCount} 个，失败 {failedCount} 个；{names}{remaining} 已保留原始日志；源 ZIP 已保留",
                false);
        }
        catch (Exception exception)
        {
            if (stagingDirectory is not null)
            {
                if (publicationTracker is not null)
                    await publicationTracker.CancelPublicationAsync(stagingDirectory, CancellationToken.None).ConfigureAwait(false);
                TryDeleteDirectory(stagingDirectory);
            }
            return new DecodeResult(
                request, DecodeState.Failed, null,
                $"{exception.Message}；未生成输出目录，源 ZIP 已保留", false);
        }
    }

    private async Task<IReadOnlyList<ArchiveItem>> InspectEntriesAsync(
        ZipArchive archive, string sourcePath, CancellationToken cancellationToken)
    {
        if (archive.Entries.Count > MaximumArchiveEntryCount)
            throw DecodeException.DecodingFailed($"ZIP 条目总数超过 {MaximumArchiveEntryCount} 个");
        var items = new List<ArchiveItem>();
        var seen = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        long totalSize = 0;
        foreach (var entry in archive.Entries)
        {
            cancellationToken.ThrowIfCancellationRequested();
            var relativePath = ZipPathValidator.ValidateRelativePath(entry.FullName);
            if (!seen.Add(relativePath))
                throw DecodeException.DecodingFailed($"ZIP 包含重复路径：{relativePath}");
            if (ZipPathValidator.IsMetadataPath(relativePath)) continue;
            var kind = EntryKind(entry);
            if (kind == ArchiveItemKind.Special)
                throw DecodeException.DecodingFailed($"ZIP 包含符号链接或特殊条目：{relativePath}");
            if (kind == ArchiveItemKind.File)
            {
                if (entry.Length > MaximumEntrySize)
                    throw DecodeException.DecodingFailed($"ZIP 条目超过 500 MB：{relativePath}");
                DecodeLimits.ValidateOutputSize(totalSize, entry.Length, MaximumTotalSize);
                totalSize += entry.Length;
            }
            LogFormat? format = null;
            if (kind == ArchiveItemKind.File)
            {
                var virtualPath = Path.Combine(
                    Path.GetDirectoryName(sourcePath) ?? "",
                    relativePath.Replace('/', Path.DirectorySeparatorChar));
                var detected = await _entryFormatResolver(virtualPath, cancellationToken).ConfigureAwait(false);
                if (detected is LogFormat.Xlog or LogFormat.Mx or LogFormat.Logan) format = detected;
            }
            items.Add(new ArchiveItem(entry, relativePath, kind, format));
        }
        return items;
    }

    private static ArchiveItemKind EntryKind(ZipArchiveEntry entry)
    {
        var unixType = (entry.ExternalAttributes >> 16) & 0xf000;
        if (unixType != 0 && unixType is not 0x8000 and not 0x4000) return ArchiveItemKind.Special;
        if ((entry.ExternalAttributes & 0x400) != 0) return ArchiveItemKind.Special;
        var directory = entry.FullName.EndsWith('/') || entry.FullName.EndsWith('\\') ||
            (entry.ExternalAttributes & (int)FileAttributes.Directory) != 0 || unixType == 0x4000;
        return directory ? ArchiveItemKind.Directory : ArchiveItemKind.File;
    }

    private static async Task ExtractPreservedItemAsync(
        ArchiveItem item, string stagingDirectory, CancellationToken cancellationToken)
    {
        var destination = Path.Combine(
            stagingDirectory, item.RelativePath.Replace('/', Path.DirectorySeparatorChar));
        if (item.Kind == ArchiveItemKind.Directory)
        {
            Directory.CreateDirectory(destination);
            return;
        }
        var payload = await ExtractDataAsync(item, cancellationToken).ConfigureAwait(false);
        await WriteExclusiveAsync(payload, destination, cancellationToken).ConfigureAwait(false);
    }

    private static async Task<byte[]> ExtractDataAsync(ArchiveItem item, CancellationToken cancellationToken)
    {
        await using var input = item.Entry.Open();
        using var output = new MemoryStream((int)Math.Min(item.Entry.Length, 1024 * 1024));
        var crc = new Crc32();
        var buffer = new byte[64 * 1024];
        while (true)
        {
            var count = await input.ReadAsync(buffer, cancellationToken).ConfigureAwait(false);
            if (count == 0) break;
            DecodeLimits.ValidateOutputSize(output.Length, count, MaximumEntrySize);
            crc.Append(buffer.AsSpan(0, count));
            output.Write(buffer, 0, count);
        }
        if (output.Length != item.Entry.Length)
            throw DecodeException.DecodingFailed($"{item.RelativePath} 解压后的大小不完整");
        if (crc.GetCurrentHashAsUInt32() != item.Entry.Crc32)
            throw DecodeException.DecodingFailed($"{item.RelativePath} ZIP CRC 校验失败");
        return output.ToArray();
    }

    private static string UniqueDecodedOutput(string relativePath, string stagingDirectory)
    {
        var relativeParent = Path.GetDirectoryName(relativePath.Replace('/', Path.DirectorySeparatorChar));
        var baseName = Path.GetFileNameWithoutExtension(relativePath);
        for (var index = 0; ; index++)
        {
            var suffix = index == 0 ? "" : $"-{index}";
            var fileName = $"{baseName}{suffix}.log";
            var candidate = relativeParent is null
                ? Path.Combine(stagingDirectory, fileName)
                : Path.Combine(stagingDirectory, relativeParent, fileName);
            if (!File.Exists(candidate) && !Directory.Exists(candidate)) return candidate;
        }
    }

    private async Task<string> MoveToUniqueOutputDirectoryAsync(
        string stagingDirectory, string sourcePath, CancellationToken cancellationToken)
    {
        var parent = Path.GetDirectoryName(sourcePath) ?? "";
        var baseName = Path.GetFileNameWithoutExtension(sourcePath);
        for (var index = 0; ; index++)
        {
            cancellationToken.ThrowIfCancellationRequested();
            var suffix = index == 0 ? "" : $"-{index}";
            var candidate = Path.Combine(parent, $"{baseName}{suffix}");
            if (publicationTracker is not null)
                await publicationTracker.PreparePublicationAsync(stagingDirectory, candidate, cancellationToken).ConfigureAwait(false);
            if (WindowsFilePublication.TryMoveExclusive(stagingDirectory, candidate)) return candidate;
        }
    }

    private static async Task WriteExclusiveAsync(
        ReadOnlyMemory<byte> data, string path, CancellationToken cancellationToken)
    {
        Directory.CreateDirectory(Path.GetDirectoryName(path) ?? "");
        await using var stream = new FileStream(
            path, FileMode.CreateNew, FileAccess.Write, FileShare.None,
            64 * 1024, FileOptions.Asynchronous | FileOptions.WriteThrough);
        await stream.WriteAsync(data, cancellationToken).ConfigureAwait(false);
        await stream.FlushAsync(cancellationToken).ConfigureAwait(false);
        stream.Flush(flushToDisk: true);
    }

    private static ValueTask<LogFormat?> DefaultEntryFormatResolver(
        string path, CancellationToken cancellationToken)
    {
        try { return ValueTask.FromResult<LogFormat?>(LogFormats.Detect(path)); }
        catch { return ValueTask.FromResult<LogFormat?>(null); }
    }

    private static void TryDelete(string? path)
    {
        try { if (path is not null && File.Exists(path)) File.Delete(path); }
        catch { }
    }

    private static void TryDeleteDirectory(string path)
    {
        try { if (Directory.Exists(path)) Directory.Delete(path, recursive: true); }
        catch { }
    }

    private enum ArchiveItemKind { File, Directory, Special }
    private sealed record ArchiveItem(
        ZipArchiveEntry Entry, string RelativePath, ArchiveItemKind Kind, LogFormat? Format);
}
